controllers/eventbus/installer/jetstream.go

```diff
@@ -0,0 +1,615 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: r.eventBus.Namespace,
+			Name:      generateJetStreamServiceName(r.eventBus),
+			Labels:    r.labels,
+			Annotations: map[string]string{
+				common.AnnotationResourceSpecHash: hash,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(r.eventBus.GetObjectMeta(), v1alpha1.SchemaGroupVersionKind),
+			},
+		},
+		Spec: spec,
+	}
+	old := &corev1.Service{}
+	if err := r.client.Get(ctx, client.ObjectKeyFromObject(obj), old); err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := r.client.Create(ctx, obj); err != nil {
+				return fmt.Errorf("failed to create jetstream service, err: %w", err)
+			}
+			r.logger.Info("created jetstream service successfully")
+			return nil
+		} else {
+			return fmt.Errorf("failed to check if jetstream service is existing, err: %w", err)
+		}
+	}
+	if old.GetAnnotations()[common.AnnotationResourceSpecHash] != hash {
+		old.Annotations[common.AnnotationResourceSpecHash] = hash
+		old.Spec = spec
+		if err := r.client.Update(ctx, old); err != nil {
+			return fmt.Errorf("failed to update jetstream service, err: %w", err)
+		}
+		r.logger.Info("updated jetstream service successfully")
+	}
+	return nil
+}
+
+func (r *jetStreamInstaller) createStatefulSet(ctx context.Context) error {
+	jsVersion, err := r.config.GetJetStreamVersion(r.eventBus.Spec.JetStream.Version)
+	if err != nil {
+		return fmt.Errorf("failed to get jetstream version, err: %w", err)
+	}
+	spec := r.buildStatefulSetSpec(jsVersion)
+	hash := common.MustHash(spec)
+	obj := &appv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: r.eventBus.Namespace,
+			Name:      generateJetStreamStatefulSetName(r.eventBus),
+			Labels:    r.labels,
+			Annotations: map[string]string{
+				common.AnnotationResourceSpecHash: hash,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(r.eventBus.GetObjectMeta(), v1alpha1.SchemaGroupVersionKind),
+			},
+		},
+		Spec: spec,
+	}
+	old := &appv1.StatefulSet{}
+	if err := r.client.Get(ctx, client.ObjectKeyFromObject(obj), old); err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := r.client.Create(ctx, obj); err != nil {
+				return fmt.Errorf("failed to create jetstream statefulset, err: %w", err)
+			}
+			r.logger.Info("created jetstream statefulset successfully")
+			return nil
+		} else {
+			return fmt.Errorf("failed to check if jetstream statefulset is existing, err: %w", err)
+		}
+	}
+	if old.GetAnnotations()[common.AnnotationResourceSpecHash] != hash {
+		old.Annotations[common.AnnotationResourceSpecHash] = hash
+		old.Spec = spec
+		if err := r.client.Update(ctx, old); err != nil {
+			return fmt.Errorf("failed to update jetstream statefulset, err: %w", err)
+		}
+		r.logger.Info("updated jetstream statefulset successfully")
+	}
+	return nil
+}
+
+func (r *jetStreamInstaller) buildStatefulSetSpec(jeVersion *controllers.JetStreamVersion) appv1.StatefulSetSpec {
```

Comment:
Having a default version will cause future upgrade issue - if `version` is not specified, we use a default version, that will prevent us from updating the default version later, otherwise if the default one is changed, those installation with default versions will be automatically updated. This is a problem for current nats streaming implementation.

Reply:
Got it, thanks for the clarification